example/kokkos_pi/kokkos_pi.cpp

```diff
@@ -0,0 +1,381 @@
+                                           Kokkos::ChunkSize(512));
+  double theSum = 0.0;
+  // reduce the sum on the device
+  // I'm pretty sure I can do this better, but not worried about performance for this
+  MeshBlock *pmb = firstBlock;
+  while (pmb) {
+    Container<Real> &base = pmb->real_containers.Get();
+    auto inOrOut = base.PackVariables({Metadata::Independent});
+    double oneSum;
+    Kokkos::parallel_reduce(
+        "Reduce Sum", policyBlock,
+        KOKKOS_LAMBDA(const int &idx, double &mySum) {
+          const int k_grid = idx / n_block2;
+          const int j_grid = (idx - k_grid * n_block2) / n_block;
+          const int i_grid = idx - k_grid * n_block2 - j_grid * n_block;
+          mySum += inOrOut(0, k_grid + NGHOST, j_grid + NGHOST, i_grid + NGHOST);
+        },
+        oneSum);
+    Kokkos::fence();
+    theSum += oneSum;
+    pmb = pmb->next;
+  }
+  // calculate Pi
+  return theSum;
+}
+
+static MeshBlock *setupMesh(const int &n_block, const int &n_mesh, const double &radius,
+                            View2D &xyz, const int NG = 0) {
+  // *** Kludge warning ***
+  // Since our mesh is not GPU friendly we set up a hacked up
+  // collection of mesh blocks.  The hope is that when our mesh is
+  // up to par we will replace this code with the mesh
+  // infrastructure.
+
+  const Real dxyzCell = radius / static_cast<Real>(n_mesh * n_block);
+  auto h_xyz = Kokkos::create_mirror_view(xyz);
+
+  // Set up our mesh.
+  Metadata myMetadata({Metadata::Independent, Metadata::Cell});
+  MeshBlock *firstBlock = nullptr;
+  MeshBlock *lastBlock = nullptr;
+
+  // compute an offset due to ghost cells
+  double delta = dxyzCell * static_cast<Real>(NG);
+
+  int idx = 0; // an index into Block coordinate array
+  for (int k_mesh = 0; k_mesh < n_mesh; k_mesh++) {
+    for (int j_mesh = 0; j_mesh < n_mesh; j_mesh++) {
+      for (int i_mesh = 0; i_mesh < n_mesh; i_mesh++, idx++) {
+        // get a new meshblock and insert into chain
+        auto *pmb = new MeshBlock(n_block, 3);
+        if (lastBlock) {
+          lastBlock->next = pmb;
+          pmb->prev = lastBlock;
+        } else {
+          firstBlock = pmb;
+        }
+        // set coordinates of first cell center
+        h_xyz(0, idx) = dxyzCell * (static_cast<Real>(i_mesh * n_block) + 0.5) - delta;
+        h_xyz(1, idx) = dxyzCell * (static_cast<Real>(j_mesh * n_block) + 0.5) - delta;
+        h_xyz(2, idx) = dxyzCell * (static_cast<Real>(k_mesh * n_block) + 0.5) - delta;
+        // Add variable for in_or_out
+        Container<Real> &base = pmb->real_containers.Get();
+        base.setBlock(pmb);
+        base.Add("in_or_out", myMetadata);
+        // repoint lastBlock for next iteration
+        lastBlock = pmb;
+      }
+    }
+  }
+  // copy our coordinates over to Device and wait for completion
+  Kokkos::deep_copy(xyz, h_xyz);
+  Kokkos::fence();
+
+  return firstBlock;
+}
```

Comment:
This isn't necessary, but it might be interesting:

If meshblocks can be copied to the GPU via capture, perhaps we could make a Kokkos view of them to represent a mesh rather than a tree. The advantage of such a test would be if coalescing over meshblocks offers a big performance boost.

I think for this code, that would be a different setup routine, like `setupKokkosMesh` or something.

Reply:
I'll try my hand at that next.  I was also thinking of playing with nested parallelism of Kokkos since we can create a parND view-of-views across multiple blocks and hand that off to the par_for.

Reply:
Yeah, that's a good idea.

One thing that would certainly work would be a (yuck) Kokkos view of VariablePacks, which would also achieve what we want. A `VariableBlockPack` or something.